File: medium/Subsets_II.py
class Solution:
    def __init__(self):
        self.answer = [[]]
    
    def next_number(self, nums, answer, cnt):
        if len(answer) == cnt:
            self.answer.append(answer)
        else:
            for num in set(nums):
                index = nums.index(num)
                self.next_number(nums[index+1:], answer+[num], cnt)
    
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        for r in range(1, len(nums)):
            for num in set(nums):
                index = nums.index(num)
                self.next_number(nums[index+1:], [num], r)
        self.answer.append(nums)
        return self.answer

File: cpp/medium/Subsets_II.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> answer;
        vector<int> combination;
        sort(nums.begin(), nums.end());
        dfs(nums, combination, answer);
        return answer;
    }
    void dfs(vector<int>& nums, vector<int>& combination, vector<vector<int>>& answer) {
        answer.push_back(combination);
        set<int> set_num(nums.begin(), nums.end());
        for (auto &i:set_num){
            combination.push_back(i);
            vector<int> sub_num(find(nums.begin(), nums.end(), i)+1, nums.end());
            dfs(sub_num, combination, answer);
            combination.pop_back();
        }
    }
};
